4 allow the user to edit items

tycebrown opened the pull request:
Got the functionality to work, but my method was a little... sketchy. Would love to hear your input.

tycebrown committed:
Made it possible for the user to edit items

1. encapsulated each item into a class (Item)
2. added the item widget class.
3. ...and finally, added the ability for the user to edit items.

tycebrown committed:
Merge branch 'main' into 4-allow-the-user-to-edit-items

tycebrown committed:
Added disposal functionality of the ItemWidget and Item

Pertempto commented on lib/item.dart:
```diff
@@ -0,0 +1,38 @@
+  //     ItemWidget, constantly updating the value of its
+  //     corresponding item. I think this would be redundant and
+  //     add additional complexity.
+  // However, my solution seems to have a "code smell." What do you
```
You are correct, this isn't a good design pattern. This class is meant for storing data, but it contains a widget.  The instances of this class are part of the "model" of your app. You don't want to store widgets in an object that is part of the model. Widgets are part of the "view" of your app. You want to keep the model and view separated where you can.  Skim over [this article](https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller) for some background.

Pertempto replied:
The most important thing to do is to remove the `TextEditingController` from the `Item` class. I left two possible solutions below.

tycebrown replied:
Oh, that makes sense.

Pertempto commented on lib/item_widget.dart:
```diff
@@ -0,0 +1,36 @@
+    final itemTextStyle = Theme.of(context).textTheme.titleLarge;
+
+    return ListTile(
+      title: TextFormField(
```
I'm not sure this is the best design. It looks strange to have a list of `TextFormField` on the screen.   

This is the way I was envisioning it:  
* I would recommend replacing the `TextFormField` with `Text`
* Then add an `onTap` callback to the `ListTile`.  
* When the user taps a `ListTile`, open an `AlertDialog` that contains the `TextFormField`.
* The user enters the updated text into the dialog, and then presses some kind of confirmation button.
* When the confirmation button is tapped, the `Item` is updated with the new text.

Pertempto replied:
Adding a dialog does have the drawback that it adds an extra tap for the user, to confirm the dialog. Your current solution allows the user to jump from editing one item to another quickly, which is very nice. You could totally keep your current implementation, but you need to move the `TextEditingController` to `_ItemWidgetState`.  
* Use the `initialValue` parameter of `TextFormField` to set the initial value from `widget.item`.
* Use the `onChanged` parameter of `TextFormField` to get the updated value and store in `widget.item`.

Pertempto commented on lib/item.dart:
```diff
@@ -0,0 +1,38 @@
+import 'package:flutter/material.dart';
+
```
I would recommend adding a directory called `data` or `model` for containing your model files.

Pertempto commented on lib/item_widget.dart:
```diff
@@ -0,0 +1,36 @@
+import 'package:flutter/material.dart';
+import 'package:lists/item.dart';
+
```
I would recommend creating a directory called `widgets` or `view` to hold your widget files.

Pertempto reviewed (changes requested):
The functionality works well, but the code needs reworked to follow better design principles.

tycebrown committed:
Redesigned the editing functionality

I found your explanation very helpful! I took your suggestions and decided to go the path of a dialog pop-up.

Up for further review!

tycebrown commented:
I found your explanation very helpful! I took your suggestions and decided to go the path of a dialog pop-up.

Up for further review!

Pertempto commented on lib/view/item_widget.dart:
```diff
@@ -0,0 +1,65 @@
+    showDialog(
+      context: context,
+      builder: (context) => AlertDialog(
+          title: _buildEditDialogTitle(context),
+          content: _buildEditDialogContent(context),
+          actions: [_buildSubmitButton(context)]),
```
It's not really necessary to have all three of these extra `_build...` methods. Just include that code in this method.

Pertempto replied:
@tycebrown, you missed this one thing. You can remove the three extra methods. Having them as separate methods just adds unnecessary lines of code.  
I really do not like unnecessary lines of code 😁

Pertempto commented on lib/view/item_widget.dart:
```diff
@@ -0,0 +1,65 @@
+}
+
+class _ItemWidgetState extends State<ItemWidget> {
+  // I think it is better to have a the TextEditingController in the
+  // State rather than the widget (since the user of the widget doesn't
+  // need to access the controller), but I am open to other views.
```
Yes, this is correct. Please remove the comment now.

Pertempto commented on lib/view/item_widget.dart:
```diff
@@ -0,0 +1,65 @@
+    return TextFormField(
+      controller: _textEditingController,
+      autofocus: true,
+      // Note: I know there is some duplication here, since the user
+      // can submit an item both through the submit action button and
+      // pressing enter in the text form field. Is this duplication useful.
```
That's not duplication, that's a bonus feature! 😄 I like it!  
Note: these kind of comments (where you are asking about the design) shouldn't really go in the code. Post them in the pull request instead. Now you need to remove it from the code, because we don't want it in there long term.

Pertempto commented on lib/model/item.dart:
```diff
@@ -0,0 +1,12 @@
+/// Item:
+///   - objects of this class store a single item
+///   in a list (needs to be wrapped in an ItemWidget
+///   to be displayed)
```
```suggestion
///     in a list (needs to be wrapped in an ItemWidget
///     to be displayed)
```
It'll look better if the beginning of these lines line up with line #2.

Pertempto commented on lib/view/item_widget.dart:
```diff
@@ -0,0 +1,65 @@
+
+/// ItemWidget:
+///   - a widget representing the view of a single item
+///   in a list (see Item)
```
```suggestion
///     in a list (see Item)
```

Pertempto commented:
> I found your explanation very helpful! I took your suggestions and decided to go the path of a dialog pop-up.
> 
> Up for further review!

Don't forget to remove the "changes requested" label!

Pertempto reviewed (changes requested):
The functionality is working great! Just a few more items to cleanup.

tycebrown committed:
Fixed Styling and learned not to put design questions in comments.

This time, I'm removing the changes requested label :)

tycebrown commented:
This time, I removed the changes requested label :)

Thank you for being patient with me.

tycebrown committed:
Removed the extra build methods this time :)

tycebrown commented:
Oh! I think I got everything :)

Pertempto reviewed (approved):
Great work!
